Support Schematic Specification v3

gabizou opened the pull request:
**SpongeAPI** | [Sponge](https://github.com/SpongePowered/Sponge/pull/3300) | [Specification](https://github.com/SpongePowered/Schematic-Specification/pull/20)

DO NOT MERGE

Accompanying changes for enabling Schematic supporting the Spec. Some of the changes are necessitated to represent a Schematic from deserialization.

I'll jot some mental notes here as time permits, but there's some ramifications that arise from having Biomes being Server scoped. I can't imagine that the game scope would change before 1.18, but in effect, biomes are re-created every server instance, so it's very well possible that a Schematic cannot exist between server restarts (like on the client going between single player worlds).

gabizou committed:
Expand a bit of Schematic API to support serialiation

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Refactor Mutability of Volumes

An intention of having Mutable Volumes recyclable and reusable
is an effort to have them simplistic with their use and retain
their extendability. Effectively, the usages of BiomeVolume.Mutable
etc will be limited to their scope, while the other volume types
can extend Modifiable entirely.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou commented on src/main/java/org/spongepowered/api/command/parameter/Parameter.java:
```diff
@@ -713,6 +715,32 @@ static Parameter seq(@NonNull final Iterable<Parameter> parameters) {
     }
 
+    @SafeVarargs
+    static Function<CommandContext, RegistryHolder> registryFromOptions(final Function<CommandContext, Optional<RegistryHolder>>... functions){
```
@dualspiral I still don't know how we want to have this documented, but there's the use case that we already discussed before with the testplugin.

dualspiral replied:
I actually need to think about where that should go full stop - I always said here was just a temporary home after all.

dualspiral replied:
I've handled this by splitting apart getting the holder (of which multiple can be specified, choosing the first one it comes across) and `RegistryType` to get.

For reference (some of these already existed)

https://github.com/SpongePowered/SpongeAPI/blob/a5b9d9fb5dd5b84a07a8557e9c2348788112b05a/src/main/java/org/spongepowered/api/command/parameter/managed/standard/VariableValueParameters.java#L231-L263

for use with 

https://github.com/SpongePowered/SpongeAPI/blob/a5b9d9fb5dd5b84a07a8557e9c2348788112b05a/src/main/java/org/spongepowered/api/command/parameter/Parameter.java#L750-L761

gabizou commented on src/main/java/org/spongepowered/api/world/schematic/PaletteTypes.java:
```diff
@@ -58,6 +59,8 @@
      */
     public static final DefaultedRegistryReference<PaletteType<BlockState, BlockType>> BLOCK_STATE_PALETTE = PaletteTypes.key(ResourceKey.sponge("block_state_palette"));
 
+    public static final DefaultedRegistryReference<PaletteType<BlockEntityType, BlockEntityType>> BLOCK_ENTITY_PALETTE = PaletteTypes.key(ResourceKey.sponge("block_entity_type_palette"));
```
Needs to be removed.

gabizou commented on src/main/java/org/spongepowered/api/world/volume/biome/BiomeVolumeFactory.java:
```diff
@@ -0,0 +1,48 @@
+import org.spongepowered.api.world.volume.virtual.UnrealizedBiomeVolume;
+import org.spongepowered.math.vector.Vector3i;
+
+public interface BiomeVolumeFactory {
+
+    BiomeVolume.Mutable empty(Palette<Biome, Biome> palette, RegistryReference<Biome> defaultBiome, Vector3i min, Vector3i max);
+
+    BiomeVolume.Mutable copyFromRange(BiomeVolume.Streamable<@NonNull ?> existing, Vector3i newMin, Vector3i newMax);
+
+    BiomeVolume.Mutable copy(BiomeVolume.Streamable<@NonNull ?> existing);
+
+    BiomeVolume.Immutable immutableOf(BiomeVolume.Streamable<@NonNull ?> existing);
+
+    BiomeVolume.Immutable immutableOf(BiomeVolume.Streamable<@NonNull ?> existing, Vector3i newMin, Vector3i newMax);
+
+    UnrealizedBiomeVolume.Mutable empty(Vector3i min, Vector3i max);
```
I don't know if this should exist as a separate class or whether it should be nested within BiomeVolume? @Zidane @dualspiral opinions

gabizou commented on src/main/java/org/spongepowered/api/world/volume/virtual/UnrealizedBiomeVolume.java:
```diff
@@ -0,0 +1,38 @@
+import org.spongepowered.api.world.biome.Biome;
+import org.spongepowered.api.world.volume.biome.BiomeVolume;
+
+public interface UnrealizedBiomeVolume<B extends BiomeVolume> extends Virtualized<Biome, B> {
```
This isn't necessarily made use of, but it can be potentially made use by the implementation and/or plugins for "game scoped" volumes that can exist between reloads of entire registries. I did have the feeling that Schematic/ArchetypeVolume should really make use of the "virtualization" by this to further emphasize that they can be deserialized at any given point of the game scope.

dualspiral commented on src/main/java/org/spongepowered/api/world/volume/archetype/block/entity/BlockEntityArchetypeVolume.java:
```diff
@@ -91,9 +91,7 @@
-    interface Mutable<M extends Mutable<M>> extends Streamable<M>,
-        BlockVolume.Mutable<M>,
-        MutableVolume {
+    interface Modifiable<M extends Modifiable<M>> extends Streamable<M>, BlockVolume.Modifiable<M>, MutableVolume {
```
What's the difference between "Mutable" and "Modifiable"?

dualspiral commented:
Given we're now in "soft freeze", we need to consider whether some of the changes are necessary. Most of the `Mutable` -> `Modifiable` changes seem to be ABI rather that pure API breaks, but there are a couple of functions that move - I think mostly `empty` calls on `BlockVolume` were what I saw.

If we absolutely _must_ do this then this needs to be done ASAP, but if we don't _have_ to we shouldn't now.

Zidane committed:
Remove a few straggler supplier methods left.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

gabizou committed:
Simplify Palette creation

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou commented:
@dualspiral Some things we gotta do:
- Entities and BlockEntities need their positions relative to the [0,0,0] of the volume not the offset (currently min) so their positions are purely additive to the block min of the volume itself (Since we know some of the details about the minimum on copying of the entity/blockentity -> archetype, we can do this math on copy, maybe an adjustment on the volume applicator to take it into consideration?
- It feels dirty that we have a pseudo position on an archetype, maybe we should make the stream of archetype entries instead so that position is not being stored within the object?
  -  This would resolve doing the transform on position separately from the archetype being transformed
- Pass transformation around through the transformation referents and verify their transforms (there's some problems with the inversion, it's slightly going off if you try to rotate 180 across the center)

gabizou committed:
Convert VolumeStream to use Vector3d's

There's effectively a problem with positions of blocks, they're
technically considered the corners of a dotted world, and since
transformations are performed on the double accuracy positioning, we
want to be able to retain those positions as much as possible throughout
however many transformations on a volume are made.

This is mostly to support the ability to rotate and mirror volumes in
their entirety. The implementations are already semi-functional, pending
 some off by one errors to be found and fixed.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Commit:
Add Transformation for positional transformations

Highly useful for rotations and other positional transformations across
origins. Can be applied for both block positions and floating point
positions, such as for entities.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Start improving VolumePositionTranslators

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>
